### PR TITLE
DO NOT MERGE, WIP: Carry #361: Support 'no_proxy' env variable

### PR DESCRIPTION
**Pull request by olleolleolle**
This PR, which is #361 and changes to make tests pass, is in progress :fire:

**Superseded by** #658 which does the same thing without adding the concept of "default middleware".

TODO: 

- [ ] DESIGN DECISION: decide how to get the added middleware into `default` status
- [ ] make the `default` case pass its 2 test failures due to same

```
  1) Failure:
Adapters::DefaultTest#test_proxy_auth_fail [/home/travis/build/lostisland/faraday/test/adapters/integration.rb:212]:
Faraday::ConnectionFailed expected but nothing was raised.

  2) Failure:
Adapters::DefaultTest#test_proxy [/home/travis/build/lostisland/faraday/test/adapters/integration.rb:203]:
Proxy was not able to append Via header.
Expected /:3998$/ to match nil.
```

### CODE REVIEW

File: test/adapters/integration.rb [hunk 1]
```diff
@@ -235,17 +235,26 @@ def adapter_options
...
+        builder_block =
+          if adapter == :default
+            nil
+          else
+            Proc.new do |b|
+              if options[:proxy]
+                b.request :proxy, options[:proxy]
+              else
+                b.request :proxy
+              end
```
Comment (olleolleolle) [inline test/adapters/integration.rb]: Perhaps this part will just fade away when _all_ connections will have the Request::Proxy middleware?

**Comment by somethingnew2-0** [comment]
Any updates @olleolleolle?

**Comment by olleolleolle** [comment]
The main question to resolve is how to introduce the concept of enabled-by-default middleware that the user can also disable or replace. Deciding on a design for that. (Then this middleware can be in that collection.)

Because of this practical difficulty of introducing a new concept, I've considered rewriting it all to land the NO_PROXY supper in the current location in the initialize - instead of as a middleware. 

So, the status is that I'm stalled and kind of have nobody to talk to it with.

**Comment by somethingnew2-0** [comment]
Yeah, I've been looking through this PR and the Faraday source most of today.  It looks like rewriting the proxy code as a middleware for just this one feature of NO_PROXY is a little overkill.  

However, the proxy code definitely looks cleaner as a middleware module.  I can see the problem with designing this middleware to be as enabled-by-default as it makes it a unique component and almost "unmodular" because it has to be disabled to be excluded.

**Comment by olleolleolle** [comment]
@somethingnew2-0 You nailed it.

**Comment by olleolleolle** [comment]
See also #658.

**Comment by olleolleolle** [comment]
Closing in favor of #658.

**Comment by iMacTia** [comment]
Thanks @olleolleolle for attempting this.

I agree with you and @somethingnew2-0 that having a middleware as suggested by @mislav is indeed the best thing, but I also understand this solution can be complicated to be implemented in the current gem structure.
I decided to review @somethingnew2-0 PR that address this without refactoring the proxy management into a middleware, I think it will be enough for the next release.

Thanks for your work so far, I would be happy if in the future you want to start from @somethingnew2-0 changes to create the middleware and we can discuss together the challenges you were facing while doing that. Maybe we'll need to introduce some new feature in Faraday beforehand (default middleware?)

**Comment by olleolleolle** [comment]
Thank _you_ @iMacTia and @somethingnew2-0 for great teamwork and stewardship.

Thank you to all the people involved in all the earlier steps of this effort. Props to @johnbellone who took the idea and ran with it! Thanks @mislav, for _having_ the idea! Thanks all commenters, you add tons of insight. (Hi @reggieb!)

(Does Faraday have a community chat channel?)

**Comment by iMacTia** [comment]
> (Does Faraday have a community chat channel?)
We're getting that done 😄!

**Pull request closed without merging**